src/storage-client/src/client.rs

```diff
@@ -568,6 +610,32 @@ where
             }
         }
     }
+
+    // Shared implementation for commands that install uppers with controllable
+    // behavior with encountering existing uppers.
+    fn insert_new_uppers<I: Iterator<Item = GlobalId>>(
+        &mut self,
+        ids: I,
+        skip_existing: bool,
+        // Included just for debugging in face of panic.
+        command: &StorageCommand<T>,
+    ) {
+        for id in ids {
+            if self.uppers.contains_key(&id) && skip_existing {
+                continue;
+            }
+
+            let mut frontier = MutableAntichain::new();
+            // TODO(guswynn): cluster-unification: fix this dangerous use of `as`, by
+            // merging the types that compute and storage use.
+            #[allow(clippy::as_conversions)]
+            frontier.update_iter(iter::once((T::minimum(), self.parts as i64)));
+            let part_frontiers = vec![Some(Antichain::from_elem(T::minimum())); self.parts];
+            let previous = self.uppers.insert(id, (frontier, part_frontiers));
+
+            assert!(previous.is_none(), "Protocol error: starting frontier tracking for already present identifier {:?} due to command {:?}", id, command);
```

Comment:
We have a specific error for this, `StorageError::SourceIdReused(id)`. Can you return a Result with that and assert it in the caller?

Reply:
Using the error was annoying here so I put the `GlobalId` in the error and just panic on it. lmk if you had some other vision that's more ergonomic.